### ISSUE TITLE
`tf.math.rsqrt` returns `inf` for subnormal float32 inputs on both CPU and GPU

### ISSUE THREAD

**Issue by wuyii8941** [tensorflow/tensorflow#118644]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TensorFlow 2.22.0-dev20260429

### Custom code

Yes

## Summary

`tf.math.rsqrt(x)` returns `inf` on both CPU and GPU eager for subnormal float32 inputs, where the correct result is a large but finite, representable float32 value.

For `x = 1e-40`:
- CPU eager: `inf`
- GPU eager: `inf`
- XLA: `1.0000000e+20` (correct)
- NumPy `1/sqrt`: `1.0000000e+20` (correct)

The correct result `rsqrt(1e-40) = 1e20` is well within float32 range (`max ≈ 3.4e38`).

## Scale of the problem

All subnormal float32 inputs produce `inf` on both CPU and GPU:

| x | Expected rsqrt(x) | CPU eager | GPU eager | XLA |
|---|---|---|---|---|
| `1e-40` | `1.00e+20` | `inf` | `inf` | `1.00e+20` |
| `1e-42` | `1.00e+21` | `inf` | `inf` | `1.00e+21` |
| `1e-45` | `2.98e+22` | `inf` | `inf` | `2.98e+22` |
| `tiny/4` | `5.83e+18` | `inf` | `inf` | `5.83e+18` |
| `tiny/2` | `4.12e+18` | `inf` | `inf` | `4.12e+18` |

(`tiny` = `np.finfo(np.float32).tiny` ≈ 1.175e-38)

All expected values are well within float32 range.

## Reference Behavior

NumPy returns the correct value:

```python
import numpy as np
print(1.0 / np.sqrt(np.float64(1e-40)))
# 1e+20
```

## Root Cause

The underlying issue is that `sqrt(subnormal)` returns `0.0` on both CPU and GPU:

```python
import tensorflow as tf

x = tf.constant([1e-40], dtype=tf.float32)
print(tf.math.sqrt(x).numpy())   # [0.]  — should be ~1e-20
print(tf.math.rsqrt(x).numpy())  # [inf] — because 1/0 = inf
```

`sqrt(1e-40) ≈ 1e-20`, which is itself subnormal but representable in float32. The CPU and GPU `sqrt` kernels fail to produce correct subnormal outputs, returning `0` instead. Then `rsqrt = 1/sqrt = 1/0 = inf`.

XLA handles this correctly, suggesting XLA uses a software implementation that supports subnormal intermediates.

## Expected Behavior

`tf.math.rsqrt(x)` should return a finite value when the result is representable in float32, matching XLA and NumPy.

## Actual Behavior

Both CPU and GPU eager return `inf`.

## Why This Matters

`rsqrt` is used extensively in:
- Vector normalization (`x * rsqrt(dot(x, x))`)
- Batch normalization (`rsqrt(variance + epsilon)`)
- Attention mechanisms
- Any computation involving inverse square roots of small values

When the input variance or vector magnitude is very small, `rsqrt` returning `inf` causes silent overflow in downstream computations.

## Environment

- TensorFlow 2.22.0-dev20260429
- dtype: `float32`
- CPU eager: `inf` (bug)
- GPU eager: `inf` (bug)
- XLA (`jit_compile=True`): correct finite value
- NumPy: correct finite value

## Recommendation

File in `tensorflow/tensorflow`. The root cause is in the `sqrt` kernel — both CPU and GPU implementations should handle subnormal inputs by producing correct subnormal outputs. Fixing `sqrt` would also fix `rsqrt`.

### Standalone code to reproduce the issue

```shell
## Minimal Reproduction


import tensorflow as tf

x = tf.constant([1e-40], dtype=tf.float32)

with tf.device('/CPU:0'):
    print("CPU:", tf.math.rsqrt(tf.identity(x)).numpy())
    # [inf]

with tf.device('/GPU:0'):
    print("GPU:", tf.math.rsqrt(tf.identity(x)).numpy())
    # [inf]

print("XLA:", tf.function(tf.math.rsqrt, jit_compile=True)(x).numpy())
# [1.e+20]


Observed on TensorFlow 2.22.0-dev20260429.
```

### Relevant log output

```shell

```

**Comment by dhantule** [comment]
Hi @wuyii8941, 

Thanks for reporting this and sharing the details.
We’ve received the issue and are taking a look. We’ll follow up here once we have more information or if we need any additional inputs.
Appreciate your patience.
